consensus/hotstuff/timeoutaggregator/timeout_collectors.go

```diff
@@ -0,0 +1,134 @@
+	return &TimeoutCollectors{
+		log:                log,
+		lowestRetainedView: lowestRetainedView,
+		collectors:         make(map[uint64]hotstuff.TimeoutCollector),
+		createCollector:    createCollector,
+	}
+}
+
+// GetOrCreateCollector retrieves the hotstuff.TimeoutCollector for the specified
+// view or creates one if none exists.
+//  -  (collector, true, nil) if no collector can be found by the view, and a new collector was created.
+//  -  (collector, false, nil) if the collector can be found by the view
+//  -  (nil, false, error) if running into any exception creating the timeout collector state machine
+// Expected error returns during normal operations:
+//  * mempool.DecreasingPruningHeightError - in case view is lower than lowestRetainedView
+func (t *TimeoutCollectors) GetOrCreateCollector(view uint64) (hotstuff.TimeoutCollector, bool, error) {
+	cachedCollector, hasCachedCollector, err := t.getCollector(view)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if hasCachedCollector {
+		return cachedCollector, false, nil
+	}
+
+	collector, err := t.createCollector(view)
+	if err != nil {
+		return nil, false, fmt.Errorf("could not create timeout collector for view %d: %w", view, err)
+	}
+
+	// Initial check showed that there was no collector. However, it's possible that after the
+	// initial check but before acquiring the lock to add the newly-created collector, another
+	// goroutine already added the needed collector. Hence, check again after acquiring the lock:
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	clr, found := t.collectors[view]
+	if found {
+		return clr, false, nil
+	}
+
+	t.collectors[view] = collector
+	return collector, true, nil
+}
+
+// getCollector retrieves hotstuff.TimeoutCollector from local cache in concurrent safe way.
+// Performs check for lowestRetainedView.
+// Expected error returns during normal operations:
+//  * mempool.DecreasingPruningHeightError - in case view is lower than lowestRetainedView
+func (t *TimeoutCollectors) getCollector(view uint64) (hotstuff.TimeoutCollector, bool, error) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+	if view < t.lowestRetainedView {
+		return nil, false, mempool.NewDecreasingPruningHeightErrorf("cannot retrieve collector for pruned view %d (lowest retained view %d)", view, t.lowestRetainedView)
+	}
+
+	clr, found := t.collectors[view]
+
+	return clr, found, nil
+}
+
+// PruneUpToView prunes the timeout collectors with views _below_ the given value, i.e.
+// we only retain and process whose view is equal or larger than `lowestRetainedView`.
+// If `lowestRetainedView` is smaller than the previous value, the previous value is
+// kept and the method call is a NoOp.
+func (t *TimeoutCollectors) PruneUpToView(lowestRetainedView uint64) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	if t.lowestRetainedView >= lowestRetainedView {
+		return
+	}
+	if len(t.collectors) == 0 {
+		t.lowestRetainedView = lowestRetainedView
+		return
+	}
+
+	sizeBefore := len(t.collectors)
```

Comment:
I would move the declaration of `sizeBefore` just before [line 103](https://github.com/onflow/flow-go/blob/f40669d422f7f67b193c0dc8a35e95220559fd56/consensus/hotstuff/timeoutaggregator/timeout_collectors.go#L103) and then use `sizeBefore` everywhere